tools/find_device.c

```diff
@@ -0,0 +1,146 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <libudev.h>
+#include <locale.h>
+#include <unistd.h>
+#include <stdbool.h>
+
+#include "find_device.h"
+
+static int match(char *sysname)
+{
+    struct udev *udev;
+    struct udev_enumerate *enumerate;
+    struct udev_list_entry *device;
+    struct udev_device *dev;
+    const char *path;
+    bool ret = false;
+
+    /* Create the udev object */
+    udev = udev_new();
+    if (!udev)
+    {
+        printf("Can't create udev\n");
+        exit(1);
+    }
+
+    /* Create a list of the devices with the name "ttyUSB*". */
+    enumerate = udev_enumerate_new(udev);
+    udev_enumerate_add_match_sysname(enumerate, sysname);
+    udev_enumerate_scan_devices(enumerate);
+
+    /* udev_enumerate_get_list_entry returns a list of entry,
+    but we only need the first entry, because we only cares
+    about their parent which are the same for all. */
+    device = udev_enumerate_get_list_entry(enumerate);
+
+    /* Get the filename of the /sys entry for the device
+     and create a udev_device object (dev) representing it */
+    path = udev_list_entry_get_name(device);
+    dev = udev_device_new_from_syspath(udev, path);
+
+    /* In order to get information about the
+     USB device, get the parent device with the
+     subsystem/devtype pair of "usb"/"usb_device". This will
+     be several levels up the tree, but the function will find
+     it.*/
+    dev = udev_device_get_parent_with_subsystem_devtype(
+        dev,
+        "usb",
+        "usb_device");
+    if (!dev)
+    {
+        printf("Unable to find parent usb device.");
+        exit(1);
+    }
+
+    /* From here, we can call get_sysattr_value() for each file
+     in the device's /sys entry. The strings passed into these
+     functions (idProduct, idVendor, serial, etc.) correspond
+     directly to the files in the directory which represents
+     the USB device. Note that USB strings are Unicode, UCS2
+     encoded, but the strings returned from
+     udev_device_get_sysattr_value() are UTF-8 encoded. */
+    if (!strncmp(udev_device_get_sysattr_value(dev, "manufacturer"),
+                 "Codethink", 9) &&
+        !strncmp(udev_device_get_sysattr_value(dev, "product"),
+                 "Medical Plethysmograph Device", 29))
+        ret = true;
+
+    udev_device_unref(dev);
+
+    /* Free the enumerator object */
+    udev_enumerate_unref(enumerate);
+    udev_unref(udev);
+
+    return ret;
+}
+
+/* filter function for /dev nodes match */
+static int dev_match(const struct dirent *entry)
+{
+    return !strncmp(entry->d_name, "ttyACM", 6);
+}
+
+int scan(void)
+{
+    struct dirent **namelist;
+    int n, found = 0;
+
+    n = scandir("/dev/", &namelist, dev_match, NULL);
+    if (n == -1)
+    {
+        perror("scandir");
+        exit(EXIT_FAILURE);
+    }
+
+    while (n--)
+    {
+        if (match(namelist[n]->d_name))
+        {
+            sprintf(dev_node, "/dev/%s", namelist[n]->d_name);
```

Comment:
`snprintf` is safer, because it can be passed the buffer length and the return value can be used for error handling.

Reply:
+1